dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/MemToReg.java

```diff
@@ -0,0 +1,353 @@
+        for (final Map.Entry<Local, AddressOffset> entry : matcher.assignments.entrySet()) {
+            if (replacingRegisters.containsKey(entry.getValue().base)) {
+                assert entry.getKey().getUsers().isEmpty();
+                entry.getKey().tryDelete();
+            }
+        }
+        if (loadCount + storeCount > 0) {
+            logger.debug("Removed {} loads and {} stores in function \"{}\".", loadCount, storeCount, function.getName());
+        }
+    }
+
+    private List<Type> getReplacingRegisters(Alloc allocation) {
+        if (!(allocation.getArraySize() instanceof IValue sizeExpression)) {
+            return null;
+        }
+        final int size = sizeExpression.getValueAsInt();
+        if (size != 1) {
+            //TODO arrays
+            return null;
+        }
+        final List<Type> replacementTypes = new ArrayList<>();
+        final Type type = allocation.getAllocationType();
+        //TODO PointerType
+        if (type instanceof IntegerType || type instanceof BooleanType) {
+            replacementTypes.add(type);
+        } else {
+            //TODO aggregate types
+            return null;
+        }
+        //TODO check for mixed-size accesses
+        return replacementTypes;
+    }
+
+    // Invariant: base != null
+    private record AddressOffset(RegWriter base, int offset) {
+        private AddressOffset increase(int o) {
+            return o == 0 ? this : new AddressOffset(base, offset + o);
+        }
+    }
+
+    // Invariant: register != null
+    private record RegisterOffset(Register register, int offset) {}
+
+    // Processes events in program order.
+    // Returns a label, if it is program-ordered before the current event and its symbolic state was updated.
+    private static final class Matcher implements EventVisitor<Label> {
+
+        // Current local symbolic state.  Missing values mean irreplaceable contents.
+        private final Map<Object, AddressOffset> state = new HashMap<>();
+        // Maps labels and jumps to symbolic state information.
+        private final Map<Label, Map<Object, AddressOffset>> jumps = new HashMap<>();
+        // Since it is not allowed to get the address of a register, once an address is
+        // Join over all memory information.  Initialized to all empty.  Missing entries mean irreplaceable address.
+        private final Map<RegWriter, Set<RegWriter>> reachabilityGraph = new HashMap<>();
+        // Collects candidates to be replaced.
+        private final Map<MemoryEvent, AddressOffset> accesses = new HashMap<>();
+        // Keeps track of pointer operations.  Maps assignments to the address value they must return.
+        private final Map<Local, AddressOffset> assignments = new HashMap<>();
+        private boolean dead;
+
+        @Override
+        public Label visitEvent(Event e) {
+            if (dead) {
+                return null;
+            }
+            // Publish all values passed as arguments.
```

Comment:
This code seems to assume that the parameter is a Function call (according to the comment).
If so, this assumption should be made more explicit (a comment would be sufficient).